### PR TITLE
Rng manager

### PR DESCRIPTION
**Pull request by FalseIlyu**
Would this me in line with #384 ?

- [x] wrap std::uniform_int_distribution
- [x] wrap a float distributor
- [x] thread safe
- [x] Have an option to seed with a constant for testing

Maybe there should be a requirement for the result of the distribution to be the same accross platform.

### CODE REVIEW

File: src/Resources/RNGManager.h [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+
+namespace openblack::ressources
+{
+class RNGManager
+{
+public:
+    static RNGManager& GetInstance() {
+        static RNGManager rngManager;
+        return rngManager;
+    };
```
Comment (bwrsandman) [inline src/Resources/RNGManager.h:22]: This needs to be in the cpp file

**Comment by FalseIlyu** [comment]
This should make the RNG thread safe (no concurrent access to the underlying generator) and is (at least to me) the logical extension of using a shared generator but you're going to have a cumulative performance price as more object use it.

**Comment by FalseIlyu** [comment]
I marked the task as completed as i believe the current implementation does the job. If we can agree on the general design and on the test [here](https://github.com/FalseIlyu/openblack/blob/rng_manager_ctest/test/rng_manager/test_rng_manager.cpp) i think we can mark this as ready for review.

```
Running main() from D:\openblack\vcpkg\buildtrees\gtest\src\ase-1.11.0-7d339f16d4.clean\googletest\src\gtest_main.cc
[==========] Running 2 tests from 1 test suite.
[----------] Global test environment set-up.
[----------] 2 tests from RngManager
[ RUN      ] RngManager.debug_predictable_mono
[       OK ] RngManager.debug_predictable_mono (380 ms)
[ RUN      ] RngManager.debug_predictable_multi
[       OK ] RngManager.debug_predictable_multi (1079 ms)
[----------] 2 tests from RngManager (1460 ms total)
[----------] Global test environment tear-down
[==========] 2 tests from 1 test suite ran. (1460 ms total)
[  PASSED  ] 2 tests.
```

**Comment by bwrsandman** [comment]
General style note, we try to be consistent which means `camelCase` instead of `snake_case`.
Member variables have the underscore at the end, not the start.
Also clang-format doesn't catch this but try to add a new line at the end of each file.

**Comment by bwrsandman** [comment]
I don't think resources is the right namespace, we just that namespace for resources like textures and meshes using entt's resources API.

File: src/Resources/RNGManager.h [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+ *
+ * For a complete list of all authors, please refer to contributors.md
+ * Interested in contributing? Visit https://github.com/openblack/openblack
+ *
+ * openblack is licensed under the GNU General Public License version 3.
+ *****************************************************************************/
+
+#pragma once
+
+#include <atomic>
```
Comment (bwrsandman) [inline src/Resources/RNGManager.h:12]: I don't think this include is necessary

File: src/Resources/RNGManager.h [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+#include <mutex>
+#include <random>
+
+namespace openblack::resources
+{
+class RNGManager
+{
+public:
+	static RNGManager& GetInstance();
+	int NextInt(int min = 1, int max = 500);
```
Comment (bwrsandman) [inline src/Resources/RNGManager.h:22]: I think uint32 would be better and default value agent really useful for ints

**Review by bwrsandman — CHANGES_REQUESTED** [review]

File: openblack.log [hunk 4]
```diff
@@ -0,0 +1,990 @@
+[2022-04-23 21:42:33.566] [game] [info] [Game.cpp:101] current binary path: D:/openblack/cmake-build-presets/msvc-17-vcpkg/bin/Release
```
Comment (bwrsandman) [inline openblack.log:1]: You accidentally included the log file. This should be in gitignore

File: src/Resources/RNGManager.h (hunk 3 above)
Comment (FalseIlyu) [reply src/Resources/RNGManager.h]: Should i keep allowing to define bounds on call of NextInt like in 7fca844?

**Comment by bwrsandman** [comment]
I will test with my branch and get back to you. It might take a few days.

File: src/Common/RNGManager.cpp [hunk 5]
```diff
@@ -0,0 +1,50 @@
...
+
+RNGManager& RNGManager::GetInstance()
+{
+	static RNGManager rngManager;
+	return rngManager;
+}
+
+uint32_t RNGManager::NextInt(uint32_t min, uint32_t max)
+{
+	static std::uniform_int_distribution<uint32_t> int_dist(min, max);
```
Comment (bwrsandman) [inline src/Common/RNGManager.cpp:22]: This is not thread safe and it doesn't make sense to have it thread local either.

File: src/Common/RNGManager.h [hunk 6]
```diff
@@ -0,0 +1,33 @@
...
+#include <mutex>
+#include <random>
+
+namespace openblack
+{
+class RNGManager
+{
+public:
+	static RNGManager& GetInstance();
+	uint32_t NextInt(uint32_t min, uint32_t max);
```
Comment (bwrsandman) [inline src/Common/RNGManager.h:21]: NextUint32

File: src/Common/RNGManager.cpp [hunk 7]
```diff
@@ -0,0 +1,50 @@
...
+	return float_dist(generator);
+}
+
+bool RNGManager::SetDebugMode(bool isDebug, int seed)
+{
+	std::lock_guard<std::mutex> safe_lock(_generatorLock);
+	_debugRng = true;
+	_generator.seed(seed);
+	return _debugRng == isDebug;
+}
```
Comment (bwrsandman) [inline src/Common/RNGManager.cpp:50]: Line return

File: src/Common/RNGManager.cpp [hunk 8]
```diff
@@ -0,0 +1,50 @@
...
+		std::lock_guard<std::mutex> safe_lock(_generatorLock);
+		return int_dist(_generator);
+	}
+	thread_local std::mt19937 generator;
+	return int_dist(generator);
+}
+
+float RNGManager::NextFloat(float min, float max)
+{
+	static std::uniform_real_distribution<float> float_dist(min, max);
```
Comment (bwrsandman) [inline src/Common/RNGManager.cpp:34]: Not thread safe.

File: src/Common/RNGManager.h [hunk 9]
```diff
@@ -0,0 +1,33 @@
...
+	float NextFloat(float min = 0.0, float max = 1.0);
+	bool SetDebugMode(bool is_debug, int seed);
+
+private:
+	std::mt19937 _generator;
+	std::mutex _generatorLock;
+	bool _debugRng = false;
+	RNGManager() {};
```
Comment (bwrsandman) [inline src/Common/RNGManager.h:29]: The definition should be in the implementation file (*.cpp)

File: src/Common/RNGManager.h [hunk 10]
```diff
@@ -0,0 +1,33 @@
...
+	RNGManager(const RNGManager&) = delete;
+	RNGManager& operator=(const RNGManager&) = delete;
```
Comment (bwrsandman) [inline src/Common/RNGManager.h:31]: Deletes should have public visibility

File: src/Common/RNGManager.h [hunk 11]
```diff
@@ -0,0 +1,33 @@
...
+public:
+	static RNGManager& GetInstance();
+	uint32_t NextInt(uint32_t min, uint32_t max);
+	float NextFloat(float min = 0.0, float max = 1.0);
+	bool SetDebugMode(bool is_debug, int seed);
+
+private:
+	std::mt19937 _generator;
+	std::mutex _generatorLock;
+	bool _debugRng = false;
```
Comment (bwrsandman) [inline src/Common/RNGManager.h:28]: Please define in the constructor you have

File: src/Common/RNGManager.h [hunk 12]
```diff
@@ -0,0 +1,33 @@
...
+
+#include <mutex>
+#include <random>
+
+namespace openblack
+{
+class RNGManager
+{
+public:
+	static RNGManager& GetInstance();
```
Comment (bwrsandman) [inline src/Common/RNGManager.h:20]: In order to be consistent please name this `instance()`

File: src/Common/RNGManager.h [hunk 13]
```diff
@@ -0,0 +1,33 @@
...
+	uint32_t NextInt(uint32_t min, uint32_t max);
+	float NextFloat(float min = 0.0, float max = 1.0);
```
Comment (bwrsandman) [inline src/Common/RNGManager.h:22]: Not really useful to have default values.

**Review by bwrsandman — CHANGES_REQUESTED** [review]

**Comment by bwrsandman** [comment]
I think you're better off merging NextInt and NextFloat into a template.
Since you're already including `mutex` and `random` in the header, I'm not worried about added include leaks

File: .gitignore [hunk 14]
```diff
@@ -2,6 +2,9 @@
 .idea/
 cmake-build-*/
 
+#Log files
+*.log
+
```
Comment (bwrsandman) [inline .gitignore:7]: You can remove this change now

File: src/Common/RNGManager.h [hunk 15]
```diff
@@ -0,0 +1,45 @@
...
+		    std::conditional_t<std::is_integral_v<T>, std::uniform_int_distribution<T>, std::uniform_real_distribution<T>>;
+		dist_t dist(min, max);
+		if (_debugRng)
+		{
+			std::lock_guard<std::mutex> safe_lock(_generatorLock);
+			return dist(_generator);
+		}
+		thread_local std::mt19937 generator;
+		return dist(generator);
+	};
```
Comment (bwrsandman) [inline src/Common/RNGManager.h:34]: Extra `;`

**Comment by bwrsandman** [comment]
I've tested it and it works really well and exactly how I imagined. Good work!

**Review by bwrsandman — APPROVED** [review]

**Pull request merged**